core/commonMain/src/kotlinx/serialization/internal/Tagged.kt

```diff
@@ -286,7 +286,7 @@ public abstract class TaggedDecoder<Tag : Any?> : Decoder, CompositeDecoder {
         previousValue: T?
     ): T? =
         tagBlock(descriptor.getTag(index)) {
-            if (decodeNotNullMark()) decodeSerializableValue(
+            if (decodeNotNullMark() || deserializer.descriptor.isNullable) decodeSerializableValue(
                 deserializer,
                 previousValue
             ) else decodeNull()
```

Comment:
This is not correct:
1. A correct nullable serializer must call `decodeNotNullMark`  and `decodeNull` itself. It can not be called here as well in the case you want the serializer to handle the nullability.
2. There is an assumption as to the behaviour of `decodeSerializableValue`, the right way would be to actually invoke the deserializer directly (so it can do its own - maybe custom - behaviour).
3. The correct (full) implementation is provided by `decodeNullableSerializableValue()`. The code block should be replaced by that, although there is a case for adding the overload that takes the previous value (the same way `decodeSerializableValue` is overloaded).

Reply:
Unresolve if my last commit doesn't cover all if your points, but I think I got them all.